File: LeetCode-Java/p3.java
import java.util.HashMap;
import java.util.HashSet;

class Solution {
    public int lengthOfLongestSubstring(String s) {
        HashSet<Character> window = new HashSet<>();
        int l = 0;
        int res = 0;
        for (int i = 0; i < s.length(); i++) {
            while (window.contains(s.charAt(i))) {
                window.remove(s.charAt(l));
                l++;
            }
            window.add(s.charAt(i));
            res = Math.max(res, i - l + 1);
        }
        return res;
    }
}

class Solution3 {
    public int lengthOfLongestSubstring(String s) {
        if (s.length() <= 1) {
            return s.length();
        }

        HashMap<Character, Integer> di = new HashMap<>();
        int maxLength = 0;
        int start = 0;

        for (int i = 0; i < s.length(); i++) {
            if (di.containsKey(s.charAt(i)) && start <= di.get(s.charAt(i))) {
                start = di.get(s.charAt(i)) + 1;
            } else {
                int length = i - start + 1;
                maxLength = Math.max(length, maxLength);
            }
            di.put(s.charAt(i), i);
        }
        return maxLength;
    }
}

class Solution2 {
    public int lengthOfLongestSubstring(String s) {
        int ans = 0;
        for (int i = 0; i < s.length(); i++) {
            for (int j = i; j < s.length(); j++) {
                if (allUnique(s, i, j)) {
                    ans = Math.max(ans, j - i);
                }
            }
        }
        return ans;
    }

    private boolean allUnique(String s, int start, int end) {
        HashSet<Character> values = new HashSet<>();
        for (int i = start; i < end; i++) {
            char c = s.charAt(i);
            if (values.contains(c)) {
                return false;
            }
            values.add(c);
        }
        return true;
    }
}

File: LeetCode/p3.py
# Sliding Window
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        window = set()
        l = 0
        res = 0
        for i in range(len(s)):
            while s[i] in window:
                window.remove(s[l])
                l += 1

            window.add(s[i])
            res = max(res, i - l + 1)

        return res


# 3. Longest Substring Without Repeating Characters3. Longest Substring Without Repeating Characters
class Solution3:
    def lengthOfLongestSubstring(self, s: str) -> int:

        # check length first
        if len(s) <= 1:
            return len(s)

        # create a dict to track records or not
        di = {}
        # this will hold max length
        max_length = 0
        # create an start point
        start = 0

        for i in range(len(s)): # O(n)
            if s[i] in di and start <= di[s[i]]: #O(1)
                # that means, it's already seen
                # so increase the start
                start = di[s[i]] + 1

            else:
                length = i - start + 1
                # get the max value
                max_length = max(length, max_length)

            # save current character into dict
            di[s[i]] = i

        return max_length

    # complexity: O(n)


# Approach 1: Brute Force
class Solution2:
    def lengthOfLongestSubstring(self, s: str) -> int:
        ans = 0
        for i in range(len(s)):  # o(n)
            for j in range(i, len(s)):  # o(n)
                if self.allUnique(s, i, j):  # o(n) all unique use loop
                    ans = max(ans, j-1)

        return ans

    def allUnique(self, sub: str, start: int, end: int) -> bool:
        values = set()
        for i in range(start, end):
            char = sub[i]

            if char in values:
                return False

            set.add(char)

        return True

    # total complexity O(N) * O(N) * O(N) = O(N^3)
